text/0000-contract-metadata.md

```diff
@@ -0,0 +1,117 @@
+- Proposal Name: `contract_metadata`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- NEP PR: [nearprotocol/neps#0003](https://github.com/nearprotocol/NEPs/pull/3)
+
+# Summary
+[summary]: #summary
+
+This NEP introduces contract metadata, which summarizes the content of a given contract in json format.
+Contract metadata allows developers to easily list the methods in a contract and potentially display them to users to
+provide transparency.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no direct way for a developer to programmatically list the methods of a contract that is deployed on chain
```

Comment:
Technically, this is not correct. If names were mangled host wouldn't be able to call the method from Wasm. The argument names might be mangled though.

Reply:
Are the names stored in a table or are they in a header file, or something else? If the names and types and offsets of all functions are programmatically available already, the metadata wouldn't need any of that?

Reply:
I leave this research to @bowenwang1996 . I just know for a fact that names are not mangled, because host calls functions in Wasm by their names.

Reply:
I don't think the names are stored anywhere. Only the function names are available. Also because wasm only has i32, i64, f32, and f64, all the type information is erased.

Reply:
We should figure out it for a fact, for instance, find it in the Wasm spec, or ask it from someone developing Wasm or Wasm-related infra (you can ask Max).

>Also because wasm only has i32, i64, f32, and f64, all the type information is erased.

I don't see how this:
>Also because wasm only has i32, i64, f32, and f64

implies this:
>all the type information is erased.

I suggest we do not do own our deduction, and either find it in the Wasm spec or talk to the Wasm team.

Reply:
http://webassembly.github.io/spec/core/text/types.html

Reply:
Can you elaborate on this link?

Reply:
Also, did you look into how Wasm preserves the names of the functions? I am pretty sure they are not mangled, otherwise we would not be able to call functions by name from Wasmer.

Reply:
The link says the parameter of a function takes valtype, which is one of `i32`, `i64`, `f32`, `f64`.

Reply:
Cool, I see now that argument names are not preserved. Thank you. Could you do now the research on how function names are handled in Wasm?